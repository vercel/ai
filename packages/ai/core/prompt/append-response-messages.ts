import {
  extractMaxToolInvocationStep,
  FileUIPart,
  Message,
  ReasoningUIPart,
  TextUIPart,
  ToolInvocation,
  ToolInvocationUIPart,
} from '@ai-sdk/ui-utils';
import { ResponseMessage } from '../generate-text/step-result';
import { convertDataContentToBase64String } from './data-content';
import { AISDKError } from '@ai-sdk/provider';

/**
 * Appends the ResponseMessage[] from the response to a Message[] (for useChat).
 * The messages are converted to Messages before being appended.
 * Timestamps are generated for the new messages.
 *
 * @returns A new Message[] with the response messages appended.
 */
export function appendResponseMessages({
  messages,
  responseMessages,
  _internal: { currentDate = () => new Date() } = {},
}: {
  messages: Message[];
  responseMessages: ResponseMessage[];

  /**
Internal. For test use only. May change without notice.
     */
  _internal?: {
    currentDate?: () => Date;
  };
}): Message[] {
  const clonedMessages = structuredClone(messages);

  for (const message of responseMessages) {
    const role = message.role;

    // check if the last message is an assistant message:
    const lastMessage = clonedMessages[clonedMessages.length - 1];
    const isLastMessageAssistant = lastMessage.role === 'assistant';

    switch (role) {
      case 'assistant': {
        function getToolInvocations(step: number) {
          return (
            typeof message.content === 'string'
              ? []
              : message.content.filter(part => part.type === 'tool-call')
          ).map(call => ({
            state: 'call' as const,
            step,
            args: call.args,
            toolCallId: call.toolCallId,
            toolName: call.toolName,
          }));
        }

        const parts: Array<
          TextUIPart | ReasoningUIPart | ToolInvocationUIPart | FileUIPart
        > = [];
        let textContent = '';
        let reasoningTextContent = undefined;

        if (typeof message.content === 'string') {
          textContent = message.content;
          parts.push({
            type: 'text' as const,
            text: message.content,
          });
        } else {
          let reasoningPart: ReasoningUIPart | undefined = undefined;
          for (const part of message.content) {
            switch (part.type) {
              case 'text': {
                reasoningPart = undefined; // reset the reasoning part

                textContent += part.text;
                parts.push({
                  type: 'text' as const,
                  text: part.text,
                });
                break;
              }
              case 'reasoning': {
                if (reasoningPart == null) {
                  reasoningPart = {
                    type: 'reasoning' as const,
                    reasoning: '',
                    details: [],
                  };
                  parts.push(reasoningPart);
                }

                reasoningTextContent = (reasoningTextContent ?? '') + part.text;
                reasoningPart.reasoning += part.text;
                reasoningPart.details.push({
                  type: 'text' as const,
                  text: part.text,
                  signature: part.signature,
                });
                break;
              }
              case 'redacted-reasoning': {
                if (reasoningPart == null) {
                  reasoningPart = {
                    type: 'reasoning' as const,
                    reasoning: '',
                    details: [],
                  };
                  parts.push(reasoningPart);
                }

                reasoningPart.details.push({
                  type: 'redacted' as const,
                  data: part.data,
                });
                break;
              }
              case 'tool-call':
                break;
              case 'file':
                if (part.data instanceof URL) {
                  throw new AISDKError({
                    name: 'InvalidAssistantFileData',
                    message: 'File data cannot be a URL',
                  });
                }
                parts.push({
                  type: 'file' as const,
                  mimeType: part.mimeType,
                  data: convertDataContentToBase64String(part.data),
                });
                break;
            }
          }
        }

        if (isLastMessageAssistant) {
          const maxStep = extractMaxToolInvocationStep(
            lastMessage.toolInvocations,
          );

          lastMessage.parts ??= [];

          lastMessage.content = textContent;
          lastMessage.reasoning = reasoningTextContent;
          lastMessage.parts.push(...parts);

          lastMessage.toolInvocations = [
            ...(lastMessage.toolInvocations ?? []),
            ...getToolInvocations(maxStep === undefined ? 0 : maxStep + 1),
          ];

          getToolInvocations(maxStep === undefined ? 0 : maxStep + 1)
            .map(call => ({
              type: 'tool-invocation' as const,
              toolInvocation: call,
            }))
            .forEach(part => {
              lastMessage.parts!.push(part);
            });
        } else {
          // last message was a user message, add the assistant message:
          clonedMessages.push({
            role: 'assistant',
            id: message.id,
            createdAt: currentDate(), // generate a createdAt date for the message, will be overridden by the client
            content: textContent,
            reasoning: reasoningTextContent,
            toolInvocations: getToolInvocations(0),
            parts: [
              ...parts,
              ...getToolInvocations(0).map(call => ({
                type: 'tool-invocation' as const,
                toolInvocation: call,
              })),
            ],
          });
        }

        break;
      }

      case 'tool': {
        // for tool call results, add the result to previous message:
        lastMessage.toolInvocations ??= []; // ensure the toolInvocations array exists

        if (lastMessage.role !== 'assistant') {
          throw new Error(
            `Tool result must follow an assistant message: ${lastMessage.role}`,
          );
        }

        lastMessage.parts ??= [];

        for (const contentPart of message.content) {
          // find the tool call in the previous message:
          const toolCall = lastMessage.toolInvocations.find(
            call => call.toolCallId === contentPart.toolCallId,
          );
          const toolCallPart: ToolInvocationUIPart | undefined =
            lastMessage.parts.find(
              (part): part is ToolInvocationUIPart =>
                part.type === 'tool-invocation' &&
                part.toolInvocation.toolCallId === contentPart.toolCallId,
            );

          if (!toolCall) {
            throw new Error('Tool call not found in previous message');
          }

          // add the result to the tool call:
          toolCall.state = 'result';
          const toolResult = toolCall as ToolInvocation & { state: 'result' };
          toolResult.result = contentPart.result;

          if (toolCallPart) {
            toolCallPart.toolInvocation = toolResult;
          } else {
            lastMessage.parts.push({
              type: 'tool-invocation' as const,
              toolInvocation: toolResult,
            });
          }
        }

        break;
      }

      default: {
        const _exhaustiveCheck: never = role;
        throw new Error(`Unsupported message role: ${_exhaustiveCheck}`);
      }
    }
  }

  return clonedMessages;
}
