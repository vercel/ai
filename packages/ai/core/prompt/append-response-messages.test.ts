import { describe, expect, it } from 'vitest';
import { appendResponseMessages } from './append-response-messages';

describe('appendResponseMessages', () => {
  describe('after user message', () => {
    it('appends assistant messages with text content', () => {
      const result = appendResponseMessages({
        messages: [
          {
            role: 'user',
            id: '1',
            content: 'Hello!',
            createdAt: new Date(123),
            parts: [{ type: 'text', text: 'Hello!' }],
          },
        ],
        responseMessages: [
          {
            role: 'assistant',
            content: 'This is a response from the assistant.',
            id: '123',
          },
        ],
        _internal: { currentDate: () => new Date(789) },
      });

      expect(result).toMatchSnapshot();
    });

    it('appends assistant messages with mixed complex reasoning and text', () => {
      const result = appendResponseMessages({
        messages: [
          {
            role: 'user',
            id: '1',
            content: 'Hello!',
            createdAt: new Date(123),
            parts: [{ type: 'text', text: 'Hello!' }],
          },
        ],
        responseMessages: [
          {
            role: 'assistant',
            content: [
              { type: 'reasoning', text: 'reasoning part', signature: 'sig-1' },
              { type: 'redacted-reasoning', data: 'redacted part' },
              { type: 'text', text: 'text response' },
              {
                type: 'reasoning',
                text: 'reasoning part 2',
                signature: 'sig-2',
              },
              { type: 'text', text: 'text response 2' },
            ],
            id: '123',
          },
        ],
        _internal: { currentDate: () => new Date(789) },
      });

      expect(result).toMatchSnapshot();
    });

    it('appends assistant messages with generated files', () => {
      const result = appendResponseMessages({
        messages: [
          {
            role: 'user',
            id: '1',
            content: 'Generate an image of a cat',
            createdAt: new Date(123),
            parts: [{ type: 'text', text: 'Generate an image of a cat' }],
          },
        ],
        responseMessages: [
          {
            role: 'assistant',
            content: [
              {
                type: 'file',
                mimeType: 'image/png',
                data: 'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO+ip1sAAAAASUVORK5CYII=',
              },
            ],
            id: '123',
          },
        ],
        _internal: { currentDate: () => new Date(789) },
      });

      expect(result).toMatchSnapshot();
    });

    it('handles tool calls and marks them as "call" initially', () => {
      const result = appendResponseMessages({
        messages: [
          {
            role: 'user',
            id: '1',
            content: 'User wants a tool invocation',
            createdAt: new Date(123),
            parts: [{ type: 'text', text: 'User wants a tool invocation' }],
          },
        ],
        responseMessages: [
          {
            role: 'assistant',
            content: [
              { type: 'text', text: 'Processing tool call...' },
              {
                type: 'tool-call',
                toolName: 'some-tool',
                toolCallId: 'call-1',
                args: { query: 'some query' },
              },
            ],
            id: '123',
          },
        ],
        _internal: { currentDate: () => new Date(789) },
      });

      expect(result).toMatchSnapshot();
    });

    it('adds chain of assistant messages and tool results', () => {
      const result = appendResponseMessages({
        messages: [
          {
            role: 'user',
            id: '1',
            content: 'User wants a tool invocation',
            createdAt: new Date(123),
            parts: [{ type: 'text', text: 'User wants a tool invocation' }],
          },
        ],
        responseMessages: [
          {
            role: 'assistant',
            content: [
              {
                type: 'tool-call',
                toolName: 'some-tool',
                toolCallId: 'call-1',
                args: { query: 'some query' },
              },
            ],
            id: '2',
          },
          {
            role: 'tool',
            id: '3',
            content: [
              {
                type: 'tool-result',
                toolCallId: 'call-1',
                toolName: 'some-tool',
                result: { answer: 'Tool result data' },
              },
            ],
          },
          {
            role: 'assistant',
            content: [
              {
                type: 'tool-call',
                toolName: 'some-tool',
                toolCallId: 'call-2',
                args: { query: 'another query' },
              },
            ],
            id: '4',
          },
          {
            role: 'tool',
            id: '5',
            content: [
              {
                type: 'tool-result',
                toolCallId: 'call-2',
                toolName: 'some-tool',
                result: { answer: 'another result' },
              },
            ],
          },
          {
            role: 'assistant',
            content: 'response',
            id: '6',
          },
        ],
        _internal: { currentDate: () => new Date(789) },
      });

      expect(result).toMatchSnapshot();
    });

    it('throws an error if a tool result follows a user message', () => {
      expect(() =>
        appendResponseMessages({
          messages: [
            {
              role: 'user',
              id: '1',
              content: 'User message',
              createdAt: new Date(),
              parts: [{ type: 'text', text: 'User message' }],
            },
          ],
          responseMessages: [
            {
              role: 'tool',
              id: '3',
              content: [
                {
                  type: 'tool-result',
                  toolCallId: 'call-1',
                  toolName: 'some-tool',
                  result: { answer: 'Should fail' },
                },
              ],
            },
          ],
        }),
      ).toThrowError('Tool result must follow an assistant message: user');
    });
  });

  describe('after assistant message', () => {
    it('adds assistant text response', () => {
      const result = appendResponseMessages({
        messages: [
          {
            role: 'user',
            id: '1',
            content: 'User wants a tool invocation',
            createdAt: new Date(123),
            parts: [{ type: 'text', text: 'User wants a tool invocation' }],
          },
          {
            role: 'assistant',
            id: '2',
            content: '',
            createdAt: new Date(456),
            toolInvocations: [
              {
                toolCallId: 'call-1',
                toolName: 'some-tool',
                state: 'result',
                args: { query: 'some query' },
                result: { answer: 'Tool result data' },
                step: 0,
              },
            ],
            parts: [
              {
                type: 'tool-invocation',
                toolInvocation: {
                  toolCallId: 'call-1',
                  toolName: 'some-tool',
                  state: 'result',
                  args: { query: 'some query' },
                  result: { answer: 'Tool result data' },
                  step: 0,
                },
              },
            ],
          },
        ],
        responseMessages: [
          {
            role: 'assistant',
            content: 'This is a response from the assistant.',
            id: '123',
          },
        ],
        _internal: {
          currentDate: () => new Date(789),
        },
      });

      expect(result).toMatchSnapshot();
    });

    it('adds assistant tool call response', () => {
      const result = appendResponseMessages({
        messages: [
          {
            role: 'user',
            id: '1',
            content: 'User wants a tool invocation',
            createdAt: new Date(123),
            parts: [{ type: 'text', text: 'User wants a tool invocation' }],
          },
          {
            role: 'assistant',
            id: '2',
            content: '',
            createdAt: new Date(456),
            toolInvocations: [
              {
                toolCallId: 'call-1',
                toolName: 'some-tool',
                state: 'result',
                args: { query: 'some query' },
                result: { answer: 'Tool result data' },
                step: 0,
              },
            ],
            parts: [
              {
                type: 'tool-invocation',
                toolInvocation: {
                  toolCallId: 'call-1',
                  toolName: 'some-tool',
                  state: 'result',
                  args: { query: 'some query' },
                  result: { answer: 'Tool result data' },
                  step: 0,
                },
              },
            ],
          },
        ],
        responseMessages: [
          {
            role: 'assistant',
            content: [
              {
                type: 'tool-call',
                toolName: 'some-tool',
                toolCallId: 'call-2',
                args: { query: 'another query' },
              },
            ],
            id: '123',
          },
        ],
        _internal: {
          currentDate: () => new Date(789),
        },
      });

      expect(result).toMatchSnapshot();
    });

    it('adds tool results to the previously invoked tool calls', () => {
      const result = appendResponseMessages({
        messages: [
          {
            role: 'user',
            id: '1',
            content: 'User wants a tool invocation',
            createdAt: new Date(123),
            parts: [{ type: 'text', text: 'User wants a tool invocation' }],
          },
          {
            role: 'assistant',
            id: '2',
            content: 'Placeholder text',
            createdAt: new Date(456),
            toolInvocations: [
              {
                toolCallId: 'call-1',
                toolName: 'some-tool',
                state: 'call',
                args: { query: 'some query' },
                step: 0,
              },
            ],
            parts: [
              {
                type: 'tool-invocation',
                toolInvocation: {
                  toolCallId: 'call-1',
                  toolName: 'some-tool',
                  state: 'call',
                  args: { query: 'some query' },
                  step: 0,
                },
              },
            ],
          },
        ],
        responseMessages: [
          {
            role: 'tool',
            id: '3',
            content: [
              {
                type: 'tool-result',
                toolCallId: 'call-1',
                toolName: 'some-tool',
                result: { answer: 'Tool result data' },
              },
            ],
          },
        ],
        _internal: {
          currentDate: () => new Date(789),
        },
      });

      expect(result).toMatchSnapshot();
    });

    it('appends assistant messages with mixed complex reasoning and text', () => {
      const result = appendResponseMessages({
        messages: [
          {
            role: 'user',
            id: '1',
            content: 'User wants a tool invocation',
            createdAt: new Date(123),
            parts: [{ type: 'text', text: 'User wants a tool invocation' }],
          },
          {
            role: 'assistant',
            id: '2',
            content: 'Placeholder text',
            createdAt: new Date(456),
            toolInvocations: [
              {
                toolCallId: 'call-1',
                toolName: 'some-tool',
                state: 'call',
                args: { query: 'some query' },
                step: 0,
              },
            ],
            parts: [
              {
                type: 'tool-invocation',
                toolInvocation: {
                  toolCallId: 'call-1',
                  toolName: 'some-tool',
                  state: 'call',
                  args: { query: 'some query' },
                  step: 0,
                },
              },
            ],
          },
        ],
        responseMessages: [
          {
            role: 'tool',
            id: '3',
            content: [
              {
                type: 'tool-result',
                toolCallId: 'call-1',
                toolName: 'some-tool',
                result: { answer: 'Tool result data' },
              },
            ],
          },
          {
            role: 'assistant',
            content: [
              { type: 'reasoning', text: 'reasoning part', signature: 'sig-1' },
              { type: 'redacted-reasoning', data: 'redacted part' },
              { type: 'text', text: 'text response' },
              {
                type: 'reasoning',
                text: 'reasoning part 2',
                signature: 'sig-2',
              },
              { type: 'text', text: 'text response 2' },
            ],
            id: '123',
          },
        ],
        _internal: {
          currentDate: () => new Date(789),
        },
      });

      expect(result).toMatchSnapshot();
    });
  });
});
