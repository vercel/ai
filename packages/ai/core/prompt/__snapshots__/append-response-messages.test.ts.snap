// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`appendResponseMessages > after assistant message > adds assistant text response 1`] = `
[
  {
    "content": "User wants a tool invocation",
    "createdAt": 1970-01-01T00:00:00.123Z,
    "id": "1",
    "parts": [
      {
        "text": "User wants a tool invocation",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": "This is a response from the assistant.",
    "createdAt": 1970-01-01T00:00:00.456Z,
    "id": "2",
    "parts": [
      {
        "toolInvocation": {
          "args": {
            "query": "some query",
          },
          "result": {
            "answer": "Tool result data",
          },
          "state": "result",
          "step": 0,
          "toolCallId": "call-1",
          "toolName": "some-tool",
        },
        "type": "tool-invocation",
      },
      {
        "text": "This is a response from the assistant.",
        "type": "text",
      },
    ],
    "reasoning": undefined,
    "role": "assistant",
    "toolInvocations": [
      {
        "args": {
          "query": "some query",
        },
        "result": {
          "answer": "Tool result data",
        },
        "state": "result",
        "step": 0,
        "toolCallId": "call-1",
        "toolName": "some-tool",
      },
    ],
  },
]
`;

exports[`appendResponseMessages > after assistant message > adds assistant tool call response 1`] = `
[
  {
    "content": "User wants a tool invocation",
    "createdAt": 1970-01-01T00:00:00.123Z,
    "id": "1",
    "parts": [
      {
        "text": "User wants a tool invocation",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": "",
    "createdAt": 1970-01-01T00:00:00.456Z,
    "id": "2",
    "parts": [
      {
        "toolInvocation": {
          "args": {
            "query": "some query",
          },
          "result": {
            "answer": "Tool result data",
          },
          "state": "result",
          "step": 0,
          "toolCallId": "call-1",
          "toolName": "some-tool",
        },
        "type": "tool-invocation",
      },
      {
        "toolInvocation": {
          "args": {
            "query": "another query",
          },
          "state": "call",
          "step": 1,
          "toolCallId": "call-2",
          "toolName": "some-tool",
        },
        "type": "tool-invocation",
      },
    ],
    "reasoning": undefined,
    "role": "assistant",
    "toolInvocations": [
      {
        "args": {
          "query": "some query",
        },
        "result": {
          "answer": "Tool result data",
        },
        "state": "result",
        "step": 0,
        "toolCallId": "call-1",
        "toolName": "some-tool",
      },
      {
        "args": {
          "query": "another query",
        },
        "state": "call",
        "step": 1,
        "toolCallId": "call-2",
        "toolName": "some-tool",
      },
    ],
  },
]
`;

exports[`appendResponseMessages > after assistant message > adds tool results to the previously invoked tool calls 1`] = `
[
  {
    "content": "User wants a tool invocation",
    "createdAt": 1970-01-01T00:00:00.123Z,
    "id": "1",
    "parts": [
      {
        "text": "User wants a tool invocation",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": "Placeholder text",
    "createdAt": 1970-01-01T00:00:00.456Z,
    "id": "2",
    "parts": [
      {
        "toolInvocation": {
          "args": {
            "query": "some query",
          },
          "result": {
            "answer": "Tool result data",
          },
          "state": "result",
          "step": 0,
          "toolCallId": "call-1",
          "toolName": "some-tool",
        },
        "type": "tool-invocation",
      },
    ],
    "role": "assistant",
    "toolInvocations": [
      {
        "args": {
          "query": "some query",
        },
        "result": {
          "answer": "Tool result data",
        },
        "state": "result",
        "step": 0,
        "toolCallId": "call-1",
        "toolName": "some-tool",
      },
    ],
  },
]
`;

exports[`appendResponseMessages > after assistant message > appends assistant messages with mixed complex reasoning and text 1`] = `
[
  {
    "content": "User wants a tool invocation",
    "createdAt": 1970-01-01T00:00:00.123Z,
    "id": "1",
    "parts": [
      {
        "text": "User wants a tool invocation",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": "text responsetext response 2",
    "createdAt": 1970-01-01T00:00:00.456Z,
    "id": "2",
    "parts": [
      {
        "toolInvocation": {
          "args": {
            "query": "some query",
          },
          "result": {
            "answer": "Tool result data",
          },
          "state": "result",
          "step": 0,
          "toolCallId": "call-1",
          "toolName": "some-tool",
        },
        "type": "tool-invocation",
      },
      {
        "details": [
          {
            "signature": "sig-1",
            "text": "reasoning part",
            "type": "text",
          },
          {
            "data": "redacted part",
            "type": "redacted",
          },
        ],
        "reasoning": "reasoning part",
        "type": "reasoning",
      },
      {
        "text": "text response",
        "type": "text",
      },
      {
        "details": [
          {
            "signature": "sig-2",
            "text": "reasoning part 2",
            "type": "text",
          },
        ],
        "reasoning": "reasoning part 2",
        "type": "reasoning",
      },
      {
        "text": "text response 2",
        "type": "text",
      },
    ],
    "reasoning": "reasoning partreasoning part 2",
    "role": "assistant",
    "toolInvocations": [
      {
        "args": {
          "query": "some query",
        },
        "result": {
          "answer": "Tool result data",
        },
        "state": "result",
        "step": 0,
        "toolCallId": "call-1",
        "toolName": "some-tool",
      },
    ],
  },
]
`;

exports[`appendResponseMessages > after user message > adds chain of assistant messages and tool results 1`] = `
[
  {
    "content": "User wants a tool invocation",
    "createdAt": 1970-01-01T00:00:00.123Z,
    "id": "1",
    "parts": [
      {
        "text": "User wants a tool invocation",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": "response",
    "createdAt": 1970-01-01T00:00:00.789Z,
    "id": "2",
    "parts": [
      {
        "toolInvocation": {
          "args": {
            "query": "some query",
          },
          "result": {
            "answer": "Tool result data",
          },
          "state": "result",
          "step": 0,
          "toolCallId": "call-1",
          "toolName": "some-tool",
        },
        "type": "tool-invocation",
      },
      {
        "toolInvocation": {
          "args": {
            "query": "another query",
          },
          "result": {
            "answer": "another result",
          },
          "state": "result",
          "step": 1,
          "toolCallId": "call-2",
          "toolName": "some-tool",
        },
        "type": "tool-invocation",
      },
      {
        "text": "response",
        "type": "text",
      },
    ],
    "reasoning": undefined,
    "role": "assistant",
    "toolInvocations": [
      {
        "args": {
          "query": "some query",
        },
        "result": {
          "answer": "Tool result data",
        },
        "state": "result",
        "step": 0,
        "toolCallId": "call-1",
        "toolName": "some-tool",
      },
      {
        "args": {
          "query": "another query",
        },
        "result": {
          "answer": "another result",
        },
        "state": "result",
        "step": 1,
        "toolCallId": "call-2",
        "toolName": "some-tool",
      },
    ],
  },
]
`;

exports[`appendResponseMessages > after user message > appends assistant messages with generated files 1`] = `
[
  {
    "content": "Generate an image of a cat",
    "createdAt": 1970-01-01T00:00:00.123Z,
    "id": "1",
    "parts": [
      {
        "text": "Generate an image of a cat",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": "",
    "createdAt": 1970-01-01T00:00:00.789Z,
    "id": "123",
    "parts": [
      {
        "data": "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO+ip1sAAAAASUVORK5CYII=",
        "mimeType": "image/png",
        "type": "file",
      },
    ],
    "reasoning": undefined,
    "role": "assistant",
    "toolInvocations": [],
  },
]
`;

exports[`appendResponseMessages > after user message > appends assistant messages with mixed complex reasoning and text 1`] = `
[
  {
    "content": "Hello!",
    "createdAt": 1970-01-01T00:00:00.123Z,
    "id": "1",
    "parts": [
      {
        "text": "Hello!",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": "text responsetext response 2",
    "createdAt": 1970-01-01T00:00:00.789Z,
    "id": "123",
    "parts": [
      {
        "details": [
          {
            "signature": "sig-1",
            "text": "reasoning part",
            "type": "text",
          },
          {
            "data": "redacted part",
            "type": "redacted",
          },
        ],
        "reasoning": "reasoning part",
        "type": "reasoning",
      },
      {
        "text": "text response",
        "type": "text",
      },
      {
        "details": [
          {
            "signature": "sig-2",
            "text": "reasoning part 2",
            "type": "text",
          },
        ],
        "reasoning": "reasoning part 2",
        "type": "reasoning",
      },
      {
        "text": "text response 2",
        "type": "text",
      },
    ],
    "reasoning": "reasoning partreasoning part 2",
    "role": "assistant",
    "toolInvocations": [],
  },
]
`;

exports[`appendResponseMessages > after user message > appends assistant messages with text content 1`] = `
[
  {
    "content": "Hello!",
    "createdAt": 1970-01-01T00:00:00.123Z,
    "id": "1",
    "parts": [
      {
        "text": "Hello!",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": "This is a response from the assistant.",
    "createdAt": 1970-01-01T00:00:00.789Z,
    "id": "123",
    "parts": [
      {
        "text": "This is a response from the assistant.",
        "type": "text",
      },
    ],
    "reasoning": undefined,
    "role": "assistant",
    "toolInvocations": [],
  },
]
`;

exports[`appendResponseMessages > after user message > handles tool calls and marks them as "call" initially 1`] = `
[
  {
    "content": "User wants a tool invocation",
    "createdAt": 1970-01-01T00:00:00.123Z,
    "id": "1",
    "parts": [
      {
        "text": "User wants a tool invocation",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": "Processing tool call...",
    "createdAt": 1970-01-01T00:00:00.789Z,
    "id": "123",
    "parts": [
      {
        "text": "Processing tool call...",
        "type": "text",
      },
      {
        "toolInvocation": {
          "args": {
            "query": "some query",
          },
          "state": "call",
          "step": 0,
          "toolCallId": "call-1",
          "toolName": "some-tool",
        },
        "type": "tool-invocation",
      },
    ],
    "reasoning": undefined,
    "role": "assistant",
    "toolInvocations": [
      {
        "args": {
          "query": "some query",
        },
        "state": "call",
        "step": 0,
        "toolCallId": "call-1",
        "toolName": "some-tool",
      },
    ],
  },
]
`;
